### PR TITLE
New bump_release pre-build plugin

### PR DESCRIPTION
**Pull request by twaugh**
Here is a plugin for incrementing the Release label on automatic rebuild.

The push URL can be configured separately to the fetch URL used for the build. The `GitSource` provider's repository is modified in-place, the push URL set, and 'git push' run.

For authentication, currently this is achieved with Kerberos using the kernel keyring's TGT. This works by making sure the gssproxy package is installed and having git call out to an ssh process. I haven't managed to get it to work using the pygit2 `remote.push()` method.

For other authentication mechanisms, we would need another secret to be provided. I think that requires having a service account configured, which is out of scope for this pull request.

### CODE REVIEW

File: atomic_reactor/plugins/pre_bump_release.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        if committer_name is None:
+            self.committer_name = author_name
+        if committer_email is None:
+            self.committer_email = author_email
+
+        self.push_url = push_url
+        if commit_message is not None:
+            self.commit_message = commit_message
+
+    def bump_local_file(self, parser, label_key, next_release):
```
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Wouldn't it be better to rewrite this function using dockerfile_parser? You could load the Dockerfile using that, then just traverse the loaded structure to find the proper label, change it and serialize the file again (note: it even covers cases when people use `Label` instead of `LABEL` etc).

File: atomic_reactor/plugins/pre_bump_release.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    def bump(self, repo, branch):
+        # Look in the git repository
+        remote = repo.remotes['origin']
+        if self.push_url:
+            repo.config['push.default'] = 'simple'
+            remote.push_url = self.push_url
+            remote.save()
+
+        # Bump the Release label
+        label_key = 'Release'
```
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Are we sure that this will always be `Release`? What if users provide `release` or `RELEASE`?

File: requirements.txt [hunk 3]
```diff
@@ -1,2 +1,3 @@
 docker-py
 dockerfile-parse
+#pygit2
```
Comment (bkabrda) [inline requirements.txt]: Why is this commented out? It seems that the plugin does import pygit2.

File: atomic_reactor/plugins/pre_bump_release.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        cmd = ['/usr/bin/git', 'push', 'origin']
+        try:
+            output = subprocess.check_output(cmd, stderr=subprocess.STDOUT,
+                                             cwd=repo.workdir)
+            self.log.debug("git (success): %s", output)
+        except subprocess.CalledProcessError as exc:
+            self.log.error("git (%d): %s", exc.returncode, exc.output)
+            raise
+        except AttributeError:
+            # py2.6
```
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: So why not just use the code that is in the except clause for all Python versions? IMO it works on all Python versions we care about, so I don't see the need for the try clause above.

File: atomic_reactor/plugins/pre_bump_release.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+        if self.workflow.build_process_failed:
+            self.log.info("Build already failed, not incrementing release")
+            return
+
+        # Ensure we can use the git repository already checked out for us
+        source = self.workflow.source
+        assert isinstance(source, GitSource)
+        repo = init_repository(source.get())
+        branch = repo.lookup_branch(source.git_commit)
```
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: What if user passed an actual commit instead of a branch name as `git_commit`? That can theoretically happen, AFAICS. (Isn't that what Koji will do? https://github.com/release-engineering/koji-containerbuild/blob/master/koji_containerbuild/plugins/builder_containerbuild.py#L335)

File: atomic_reactor/plugins/pre_bump_release.py (hunk 1 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: It's using DockerfileParser already. But that doesn't let us adjust a label and write it out again, so we have to reconstruct it ourselves... and we have to deal with both 'LABEL' forms.

The `'LABEL'` is used for comparing against `insn['instruction']`, which is guaranteed to be upper-case.

Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Don't know how I missed this. I guess this is ok, then.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 2 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: Good point. So, the `assert_labels` plugin defaults to requiring `Release` but is configurable: https://github.com/projectatomic/atomic-reactor/blob/master/atomic_reactor/plugins/pre_assert_labels.py.

I should probably make the name of the label configurable here.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 1 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: You missed it because the code is much larger than it looks like it ought to be when it's getting so much help. :-)

Perhaps some of this could be moved to `dockerfile_parse`.

File: requirements.txt (hunk 3 above)
Comment (twaugh) [inline requirements.txt]: There's a story here, and I forgot to tell it in the description.

Yes, in theory, we just uncomment that line. The plugin requires pygit2. However, if we do that, Travis CI will try to install it with `pip install`, and that will require building it with gcc which will fail. It will fail even if we install the dev packages it requires.

Of course, we could avoid that by simply installing the Ubuntu packaged version in the `.travis.yml` file... except pygit2 isn't packaged for Ubuntu(!) so that isn't possible after all.

So in short it's commented out because Travis CI's docker container images are based on Ubuntu.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 4 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: Fair point, will re-code.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 5 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: Yes, that's what will happen: the osbs client will be run with `--git-commit (commit-hash) --git-branch (branch-name)`. However, the client will cunningly configure the BuildConfig so that its source git ref will be the *branch* name, not the commit hash, and the `bump_release` plugin will be told the original commit hash instead.

To see how that works, take a look at the osbs-client branch tracking those changes: https://github.com/projectatomic/osbs-client/compare/master...enable-rebuild-triggers
It's in `ProductionBuild.render()`.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 1 above)
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: Filed as https://github.com/DBuildService/dockerfile-parse/issues/2.

**Comment by twaugh** [comment]
Updated with suggestions from comments.

File: atomic_reactor/plugins/pre_bump_release.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        os.environ['GIT_SSH_COMMAND'] = ssh_command
+
+        # This doesn't seem to work:
+        #   remote.push([branch.name])
+        # because it uses libssh rather than /usr/bin/ssh and so krb5
+        # auth fails.
+        # Instead, run the git command to do it
+        cmd = ['/usr/bin/git', 'push', 'origin']
+        with open('/dev/null', 'r+') as devnull:
+            p = subprocess.Popen(cmd,
```
Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: I should check whether this tries closing all open file descriptors.

Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: UPDATE: verified that it does not try to close all file descriptors, so it's fine.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 1 above)
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Great, thanks! I'm fine with keeping the code here until dockerfile-parse can help us handle it.

File: atomic_reactor/plugins/pre_bump_release.py (hunk 2 above)
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Thinking about this further, we should really expect just `Release`. Label names are case sensitive, so `Release` is a different label than `RELEASE` or `rELELASE` (just tried that).
So if someone doesn't use `Release` (precisely this way), I'd call it an error in the provided Dockerfile.

Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: OK, I'll remove the configuration option then.

File: requirements.txt (hunk 3 above)
Comment (bkabrda) [inline requirements.txt]: So I did some googling and it seems someone has solved this problem: https://github.com/uniphil/venvgit2 - it's not a very clean solution, but for tests in travis, it should be ok to use it.
I think if you add `pip install venvgit2` to `.travis.yml` before running `pip install -r requirements.txt` (with `pygit2` uncommented), it should work. Worth a try :)

Comment (twaugh) [inline requirements.txt]: Great find! Unfortunately it fails on Python 2.6: https://travis-ci.org/projectatomic/atomic-reactor/jobs/75398810

File: atomic_reactor/plugins/pre_bump_release.py (hunk 5 above)
Comment (bkabrda) [inline atomic_reactor/plugins/pre_bump_release.py]: Cunning! :)
This probably deserves a comment in code and we should also document it in osbs-client docs about build process (I added a comment to that PR that asks for this).
So IIUC, during the user-triggered build (the first one), the plugin checks that last commit in the `--git-branch` is the user-given `--git-commit`? Very nice!

Comment (twaugh) [inline atomic_reactor/plugins/pre_bump_release.py]: Yes, that's exactly it.

File: requirements.txt (hunk 3 above)
Comment (bkabrda) [inline requirements.txt]: I just filed a PR for Python 2.6 compat, so let's wait and see if upstream accepts it: https://github.com/uniphil/venvgit2/pull/12

**Comment by twaugh** [comment]
Updated with better init code.

**Comment by twaugh** [comment]
I've taken out the venvgit2 business for the time being and filed #269 for it. We can hope the upstream Python 2.6 support comes in soon.

In the mean time, this PR now passes tests.

**Comment by bkabrda** [comment]
LGTM
